### PR TITLE
BED-4537: Create Share endpoint

### PR DESCRIPTION
**Pull request by ALCooper12**
## Description

- Created a new `ShareSavedQueries` endpoint
- Added unit tests
- Updated openapi docs

## Motivation and Context

- Ticket: BED-4537
- Subtask above is from BED-4534

*Why is this change required? What problem does it solve?*
- This newly created endpoint will allow users to have the ability to share a query with others, or set it as public

## How Has This Been Tested?
- 3 successfully expected api calls via Bruno + database changes
Screenshot 2024-08-05 at 5 56 33 PM
Screenshot 2024-08-05 at 5 57 53 PM
Screenshot 2024-08-05 at 5 58 36 PM
Screenshot 2024-08-05 at 5 59 55 PM
Screenshot 2024-08-05 at 6 00 38 PM

OpenAPI preview (but please check on how this looks like for you!)
Screenshot 2024-08-05 at 6 05 56 PM

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have met the contributing prerequisites
  - Assigned myself to this PR
  - Added the appropriate labels
  - Associated an issue: https://github.com/SpecterOps/BloodHound/issues/672
  - Read the Contributing guide: https://github.com/SpecterOps/BloodHound/wiki/Contributing
- [x] I have ensured that related documentation is up-to-date
  - Open API docs
  - Code comments (GoDocs / JSDocs)
- [x] I have followed proper test practices
  - Added/updated tests to cover my changes
  - All new and existing tests passed

### CODE REVIEW

**Review by mvlipka — APPROVED** [review]
Looks good!
@irshadaj I'd still like another pair of eyes on it since I'm a bit bias here 😄

File: cmd/api/src/api/v2/saved_queries.go
Comment (mistahj67) [inline cmd/api/src/api/v2/saved_queries.go:224]: Since we are returning an array of responses anyway, why not append the errors and not return here instead?
That said, if we do want to return we have a helpful `HandleDatabaseError` util that would be perfect here
```suggestion
			// Saving query permission to one or more users
			for _, sharedUserID := range newlySharedUserIDs {
				if savedPermission, err := s.DB.CreateSavedQueryPermissionToUser(request.Context(), permissionsForSavedQuery.QueryID, sharedUserID); err != nil {
					api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusInternalServerError, api.ErrorResponseDetailsInternalServerError, request), response)
					return
				} else {
					apiResponse = append(apiResponse, savedPermission)
				}
			}
```

**Review by mistahj67 — CHANGES_REQUESTED** [review]
I had a number of questions as well that we DM'd and it looks like there's some additional work on the table still.

File: cmd/api/src/api/v2/saved_queries.go
Comment (mistahj67) [inline cmd/api/src/api/v2/saved_queries.go:224]: It looks like you can pass a slice to gorm create and it'll automatically batch it and it has an upsert / on conflict where we can just do nothing

https://gorm.io/docs/create.html#Batch-Insert
https://gorm.io/docs/create.html#Upsert-x2F-On-Conflict

File: cmd/api/src/api/v2/saved_queries.go
Comment (ALCooper12) [reply cmd/api/src/api/v2/saved_queries.go:184]: I changed it to a `api.HandleDatabaseError(request, response, err)` instead!

File: packages/go/openapi/src/schemas/model.saved-queries-permissions-request.yaml
Comment (sircodemane) [inline packages/go/openapi/src/schemas/model.saved-queries-permissions-request.yaml:22]: This field doesnt appear to exist on the request Struct, as it seems to be handled by the URL parameter instead

File: cmd/api/src/database/saved_queries_permissions.go
Comment (mistahj67) [inline cmd/api/src/database/saved_queries_permissions.go:134]: Is this missing a where clause? If we hit this with an empty id, does it wipe the entire table? Might be beneficial to toss a Limit(1) on this just in case.

File: cmd/api/src/database/saved_queries_permissions.go
Comment (mistahj67) [inline cmd/api/src/database/saved_queries_permissions.go:75]: Since we are `on conflict, do nothing` do we actually even need to do the `hasAccess` db check above? 🤔 @irshadaj 

If we do keep the check, we might consider leveraging the `GetPermissionsForSavedQuery` and checking access via logic vs hitting the db for each shared user supplied.

File: packages/go/openapi/src/schemas/model.saved-queries-permissions.yaml
Comment (sircodemane) [inline packages/go/openapi/src/schemas/model.saved-queries-permissions.yaml:24]: This property is actually covered by the model.components.int64.id part above. i suspect that this may have happened because the SavedQueryPermissions model has an ID field defined, though it would probably be better for that model to remove its ID field and replace the `Basic` embedded struct with a `BigSerial`, which has all of Basic + an int64 ID

File: .github/workflows/publish.yml
Comment (ALCooper12) [reply .github/workflows/publish.yml:1]: No not intentional 😬

File: packages/go/openapi/src/schemas/model.saved-queries-permissions.yaml
Comment (sircodemane) [inline packages/go/openapi/src/schemas/model.saved-queries-permissions.yaml:29]: This field isn't an array, it's a `uuid.NullUUID`. you can include the `null.uuid.yaml` schema here

File: cmd/api/src/api/v2/savedqueriespermissions.go
Comment (ALCooper12) [inline cmd/api/src/api/v2/savedqueriespermissions.go:186]: This was the only thing that was changed besides moving your function @mvlipka! 

It got switched from 401 to 403 since according to @mistahj67, there are some handlers on the frontend that will automatically logout a user whenever they receive a 401. And we don't want them to get logged out in this scenario

File: cmd/api/src/database/saved_queries_permissions.go
Comment (mistahj67) [inline cmd/api/src/database/saved_queries_permissions.go:104]: Lets leverage the `IsSavedQueryPublic` method here

File: cmd/api/src/database/saved_queries_permissions.go
Comment (mistahj67) [inline cmd/api/src/database/saved_queries_permissions.go:119]: Lets leverage the `IsSavedQuerySharedToUser` method here

File: cmd/api/src/database/saved_queries_permissions_test.go
Comment (mistahj67) [inline cmd/api/src/database/saved_queries_permissions_test.go:254]: Can we add the case we delete with no users and verify it deletes all

File: cmd/api/src/database/saved_queries_permissions_test.go
Comment (mistahj67) [inline cmd/api/src/database/saved_queries_permissions_test.go:47]: Can we add sharing this to a second user and then below test the scope for that user to guarantee we delete the user shared row as well?

**Review by mistahj67 — APPROVED** [review]
Looks solid! I had a few small changes to shore up testing / cleaning but don't want to block merging 🚀 once you've tackled those. Great job and thanks for your patience while we iterated through the scenarios and logic, it came out clean! 🧼

**Review by mistahj67 — APPROVED** [review]
🚀

**Pull request merged**
